File: Data Structures & Algorithms/reverse-a-linked-list/submission-7.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # initialize prev and curr pointers
        prev_node, curr_node = None, head

        # while: curr not null
        ## init. next_node – points to curr.next
        ## reverse pointer of curr – curr.next = prev
        ## prev points to curr
        ## curr points to next_node
        while curr_node is not None:
            next_node = curr_node.next
            curr_node.next = prev_node

            prev_node = curr_node
            curr_node = next_node

        # set head to prev_node
        # return head
        head = prev_node
        return head

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-7.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        // initialize slow and fast pointers – points to head node
        ListNode slow = head;
        ListNode fast = head;

        // while: (fast not null and fast.next not null)
        //// increment both pointers appropriately
        //// if: slow fast reaches slow, return true
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;

            if (fast == slow)
                return true;
        }

        // return false
        return false;
    }
}
